app/main/views/outcomes.py

```diff
@@ -0,0 +1,86 @@
+import datetime
+
+from flask import abort
+from dmapiclient.audit import AuditTypes
+
+from .. import main
+from ... import db
+from ...models import AuditEvent, Outcome
+
+from ...utils import (
+    get_json_from_request,
+    json_has_required_keys,
+    single_result_response,
+    validate_and_return_updater_request,
+)
+from ...validation import validate_outcome_json_or_400
+
+
+@main.route("/outcomes/<int:outcome_id>", methods=("GET",))
+def get_outcome(outcome_id):
+    outcome = Outcome.query.filter_by(external_id=outcome_id).first_or_404()
+
+    return single_result_response("outcome", outcome), 200
+
+
+@main.route("/outcomes/<int:outcome_id>", methods=("PUT",))
+def update_outcome(outcome_id):
+    uniform_now = datetime.datetime.utcnow()
+
+    updater_json = validate_and_return_updater_request()
+
+    json_payload = get_json_from_request()
+    json_has_required_keys(json_payload, ["outcome"])
+    outcome_json = json_payload["outcome"]
+
+    validate_outcome_json_or_400(outcome_json)
+
+    # fetch and lock Outcome row so we know writing this back won't overwrite any other updates to it that made
+    # it in in the meantime
+    outcome = db.session.query(Outcome).filter_by(
+        external_id=outcome_id,
+    ).with_for_update().first()
+
+    if not outcome:
+        abort(404, f"Outcome {outcome_id} not found")
+
+    outcome.update_from_json(outcome_json)
+
+    if outcome.completed_at is not None and outcome_json.get("completed") is False:
+        abort(400, f"Can't un-complete outcome")
+    if outcome.completed_at is None and outcome_json.get("completed") is True:
+        # make a cursory check for existing Outcome collisions. we're not actually totally relying on this
+        # to police the constraint - there is a database unique constraint that will do that for us in a transactionally
+        # bulletproof way. we do this check here manually too to be able to give a nicer error message in the 99% case.
```

Comment:
One thing we encountered during the DOS award flow was people putting in the wrong contract value (missing a decimal point and so on), and 2nd line being asked to change the value later. By changing it in the production DB 🙀 So we made the decision to allow an API user to change the contract details (they cannot change who won); the buyer can't change anything themselves from the frontend, they have to raise a support ticket. 

tl:dr - this might need to be less strict to allow for user mistakes.